docs/contributing/muzzle.md

```diff
@@ -0,0 +1,32 @@
+# Muzzle
+
+Muzzle is a feature of the Java agent that ensures API compatibility
+between libraries/symbols on the application classpath and APIs of instrumented
+3rd party libraries used by the Agent. In other words the Muzzle ensures
+that the API symbols used by the Agent are compatible with API symbols
+on the application classpath. The Muzzle will prevent loading an instrumentation
+if the APIs do not match.
+
+## How does it work
+
+At build time the Muzzle gradle plugin generates a matcher for 3rd party APIs used by the agent.
+The matcher is stored in the instrumentation class in method `ReferenceMatcher getInstrumentationMuzzle()`.
+
+At runtime the Muzzle checks API compatibility between symbols used by the Agent
+and symbols in the application class loader. Before inspecting the class loader
+the Muzzle plugin checks symbols defined in `SomeInstrumentation.typeMatcher()`
+as a performance optimization. If the symbols do not match the instrumentation
+is not loaded.
```

Comment:
another suggestion 😄

```suggestion
and symbols in the application class loader. If the symbols do not match the instrumentation is not loaded.
Because the muzzle matcher is expensive, it is only performed after a match has been made by the `classLoaderMatcher()` and `typeMatcher()` matchers. 
```

Reply:
+1. The match for `classLoaderMatcher()` and `typeMatcher()` is done by ByteBuddy right?

Reply:
Yeah, it's done a few steps before the `MuzzleMatcher` runs.